File: week_1/day-4-sliding-window/variations_que/minimum_window_substring.py
s = "ADOBECODEBANC"
t = "ABC"

def minWindow(s, t):
    from collections import Counter
    need = Counter(t)
    left = count = 0
    min_len = float('inf')
    start = 0

    for right, ch in enumerate(s):
        if ch in need:
            need[ch] -= 1
            if need[ch] >= 0:
                count += 1

        while count == len(t):
            if right - left + 1 < min_len:
                min_len = right - left + 1
                start = left

            if s[left] in need:
                need[s[left]] += 1
                if need[s[left]] > 0:
                    count -= 1
            left += 1

    return "" if min_len == float('inf') else s[start:start + min_len]

print(minWindow(s, t))  # Output: "BANC"

File: week_1/day-4-sliding-window/variations_que/minimumWindowSubstring.js
const s = "ADOBECODEBANC";
const t = "ABC";

function minWindow(s, t) {
    const need = {};
    for (let c of t) need[c] = (need[c] || 0) + 1;

    let left = 0, count = 0, minLen = Infinity, start = 0;

    for (let right = 0; right < s.length; right++) {
        if (need[s[right]] !== undefined) {
            if (--need[s[right]] >= 0) count++;
        }

        while (count === t.length) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                start = left;
            }
            if (need[s[left]] !== undefined) {
                if (++need[s[left]] > 0) count--;
            }
            left++;
        }
    }
    return minLen === Infinity ? "" : s.slice(start, start + minLen);
}

console.log(minWindow(s, t)); // BANC
